File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/XorRotate.java
package com.freight.hash;

/** xor_rotate over raw bytes. */
public final class XorRotate implements HashAlgorithm {

    @Override
    public String name() {
        return "xor_rotate";
    }

    @Override
    public long apply(byte[] data) {
        long state = 0L;
        for (int i = 0; i < data.length; i++) {
            state = (((state << 5) | (state >>> 27)) ^ (data[i] & 0xFF)) & 0xFFFFFFFFL;
        }
        return state;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/xor_rotate.cpp
#include "freight/hashes.h"

namespace freight {

// xor_rotate over raw bytes.
uint64_t hash_xor_rotate(const std::string& data) {
  uint32_t state = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    state = ((state << 5) | (state >> 27)) ^ static_cast<uint8_t>(data[i]);
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight
